scene/2d/animated_sprite_2d.cpp

```diff
@@ -275,70 +285,101 @@ void AnimatedSprite2D::_notification(int p_what) {
 }
 
 void AnimatedSprite2D::set_sprite_frames(const Ref<SpriteFrames> &p_frames) {
+	if (frames == p_frames) {
+		return;
+	}
+
 	if (frames.is_valid()) {
 		frames->disconnect(SceneStringNames::get_singleton()->changed, callable_mp(this, &AnimatedSprite2D::_res_changed));
 	}
-
+	stop();
 	frames = p_frames;
 	if (frames.is_valid()) {
 		frames->connect(SceneStringNames::get_singleton()->changed, callable_mp(this, &AnimatedSprite2D::_res_changed));
-	}
 
-	if (frames.is_null()) {
-		frame = 0;
-	} else {
-		set_frame(frame);
+		List<StringName> al;
+		frames->get_animation_list(&al);
+		if (al.size() == 0) {
+			set_animation(StringName());
+			set_autoplay(String());
+		} else {
+			if (!frames->has_animation(animation)) {
+				set_animation(al[0]);
+			}
+			if (!frames->has_animation(autoplay)) {
+				set_autoplay(String());
+			}
+		}
 	}
 
 	notify_property_list_changed();
-	_reset_timeout();
 	queue_redraw();
 	update_configuration_warnings();
+	emit_signal("sprite_frames_changed");
 }
 
 Ref<SpriteFrames> AnimatedSprite2D::get_sprite_frames() const {
 	return frames;
 }
 
 void AnimatedSprite2D::set_frame(int p_frame) {
+	set_frame_and_progress(p_frame, 0.0);
```

Comment:
If the animation plays backwards, should progress be `1.0`?

Reply:
I didn't do that because I thought it would complicate the functionality for a setter, but if you want I do that.

Reply:
fixed

Reply:
In addition, if the animation is changed during playback, the playback position is now set to the end of the animation. This applies to both AnimationSprite and AnimationPlayer.